src/main/kotlin/org/jetbrains/research/kotlin/mpp/inference/operators/layer/reccurent/lstm/LSTMLayer.kt

```diff
@@ -0,0 +1,97 @@
+import org.jetbrains.research.kotlin.mpp.inference.space.TensorRing
+import org.jetbrains.research.kotlin.mpp.inference.space.resolveSpaceWithKClass
+import org.jetbrains.research.kotlin.mpp.inference.space.toIntArray
+import org.jetbrains.research.kotlin.mpp.inference.tensors.Tensor
+import org.jetbrains.research.kotlin.mpp.inference.types.resolveKClass
+import scientifik.kmath.structures.BufferNDStructure
+import scientifik.kmath.structures.asBuffer
+import scientifik.kmath.structures.asIterable
+
+class LSTMLayer<T : Number> : RecurrentLayer<T>() {
+    @Suppress("UNCHECKED_CAST")
+    override fun apply(inputs: Collection<Tensor<T>>): Collection<Tensor<T>> {
+        require(inputs.size in 3..4) { "Applicable only for three or four arguments" }
+
+        val inputList = inputs.toList()
+
+        val inputTensor = inputList[0]
+        val weights = inputList[1].as2DCollection().first()
+        val recWeights = inputList[2].as2DCollection().first()
+        val bias = inputList.getOrNull(3)
+
+        val hiddenSize = recWeights.data.shape[1]
+        val batchSize = inputTensor.data.shape[1]
+
+        var currentState = State.initialize<T>(hiddenSize, batchSize, inputTensor.type!!)
+
+        return inputTensor.as2DCollection().map { inputMatrix ->
+            val gatesData = GatesData.create(inputMatrix, weights, recWeights, bias, currentState)
+            gatesData.activate()
+
+            val newCellGate = gatesData.forgetGate.multiply(currentState.cellGate) + gatesData.inputGate.multiply(gatesData.cellGate)
+            val newOutput = gatesData.outputGate.multiply(Activation.Tanh<T>().apply(listOf(newCellGate)).first())
+
+            currentState = State(newOutput, newCellGate)
+            newOutput.transpose()
+        }.toOutput()
+    }
+
+    data class GatesData<T : Number>(var inputGate: Tensor<T>,
+                                     var outputGate: Tensor<T>,
+                                     var forgetGate: Tensor<T>,
+                                     var cellGate: Tensor<T>) {
+        fun activate() {
+            inputGate = Activation.Sigmoid<T>().apply(listOf(inputGate)).first()
+            outputGate = Activation.Sigmoid<T>().apply(listOf(outputGate)).first()
+            forgetGate = Activation.Sigmoid<T>().apply(listOf(forgetGate)).first()
+            cellGate = Activation.Tanh<T>().apply(listOf(cellGate)).first()
+        }
+
+        companion object {
+            @Suppress("UNCHECKED_CAST")
+            fun <T : Number> create(inputMatrix: Tensor<T>, weights: Tensor<T>, recWeights: Tensor<T>,
+                                    biases: Tensor<T>?, prevState: State<T>): GatesData<T> {
+                var gates = weights.dot(inputMatrix.transpose()) + recWeights.dot(prevState.output.transpose())
+                val hiddenSize = recWeights.data.shape[1]
+                val batchSize = inputMatrix.data.shape[0]
+                if (biases != null) {
```

Comment:
Can't we make two functions -- with and without biases?

Reply:
For getting biases we using getOrNull, so we can't

Reply:
As far as I see `biases` is an argument of this function, so why can't you elevate `if` out of function and split it in two?